File: src/tetris/JPiece.java
package tetris;

import javalib.worldimages.Posn;

public class JPiece extends APiece {

	static boolean[][] JPIECE_UP = new boolean[][] {
		{ true, false, false, false }, 
		{ true, true, true, false },   
		{ false, false, false, false },
		{ false, false, false, false }
	};
	
	static boolean[][] JPIECE_LEFT = new boolean[][] {
		{ false, true, false, false }, 
		{ false, true, false, false },   
		{ true, true, false, false },
		{ false, false, false, false }
	};
	
	static boolean[][] JPIECE_RIGHT = new boolean[][] {
		{ false, true, true, false }, 
		{ false, true, false, false },   
		{ false, true, false, false },
		{ false, false, false, false }
	};
	
	static boolean[][] JPIECE_DOWN = new boolean[][] {
		{ false, false, false, false }, 
		{ true, true, true, false },   
		{ false, false, true, false },
		{ false, false, false, false }
	};
	
	
	JPiece(Posn posn) {
		super(posn, new CycleIndex(JPIECE_UP.clone(), JPIECE_LEFT.clone(), JPIECE_RIGHT.clone(), JPIECE_DOWN.clone()), Tetrimino.J);
	}

	public boolean hasSpun(Board b) {
		return false;
	}
	public boolean[][] rotInitialState(String s) {
		switch (s) {
		case "up": return JPIECE_UP;
		case "left": return JPIECE_LEFT;
		case "right": return JPIECE_RIGHT;
		default: return JPIECE_DOWN;
		}
	}

}

File: typescript/src/pieces/JPiece.ts
import { Posn } from 'impworld';
import { APiece } from './APiece';
import { CycleIndex } from './CycleIndex';
import { Tetrimino } from './Tetrimino';
import { Board } from '@/board';

export class JPiece extends APiece {
  static JPIECE_UP: boolean[][] = [
    [true, false, false, false],
    [true, true, true, false],
    [false, false, false, false],
    [false, false, false, false],
  ];
  static JPIECE_LEFT: boolean[][] = [
    [false, true, false, false],
    [false, true, false, false],
    [true, true, false, false],
    [false, false, false, false],
  ];
  static JPIECE_RIGHT: boolean[][] = [
    [false, true, true, false],
    [false, true, false, false],
    [false, true, false, false],
    [false, false, false, false],
  ];
  static JPIECE_DOWN: boolean[][] = [
    [false, false, false, false],
    [true, true, true, false],
    [false, false, true, false],
    [false, false, false, false],
  ];

  constructor(posn: Posn) {
    super(
      posn,
      new CycleIndex(
        JPiece.JPIECE_UP.slice(),
        JPiece.JPIECE_LEFT.slice(),
        JPiece.JPIECE_RIGHT.slice(),
        JPiece.JPIECE_DOWN.slice(),
      ),
      Tetrimino.J,
    );
  }

  public hasSpun(b: Board): boolean {
    return false;
  }

  public rotInitialState(s: string): boolean[][] {
    switch (s) {
      case 'up':
        return JPiece.JPIECE_UP;
      case 'left':
        return JPiece.JPIECE_LEFT;
      case 'right':
        return JPiece.JPIECE_RIGHT;
      default:
        return JPiece.JPIECE_DOWN;
    }
  }
}
